components/server/src/workspace/prebuild-rate-limiter.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+export type PrebuildRateLimiters = { [cloneURL: string]: PrebuildRateLimiterConfig } & {
+    "*": PrebuildRateLimiterConfig;
+};
+
+export type PrebuildRateLimiterConfig = {
+    // maximum number of requests per period
+    limit: number;
+
+    // time period which the limit is enforce against in seconds
+    period: number;
+};
+
+export namespace PrebuildRateLimiterConfig {
+    const DEFAULT_CONFIG: PrebuildRateLimiterConfig = {
+        limit: 50,
+        period: 50,
+    };
+
+    export function getConfigForCloneURL(
+        rateLimiters: PrebuildRateLimiters,
+        cloneURL: string,
+    ): PrebuildRateLimiterConfig {
+        // First we use any explicit overrides for a given cloneURL
+        let config = rateLimiters[cloneURL];
```

Comment:
What would be the scenario where we would want to change the rate limiting for one cloneURL but not for all?

Reply:
Note that this is the same code as before, and I'd argue out of scope for this PR.
The only change here is to pull up the time period to be configurable with changing the installer code, which is justified by the [argument made above](https://github.com/gitpod-io/gitpod/pull/14975#discussion_r1034554094).

Ultimately, I think the whole approach is ill-suited for what we want to achieve. And I feel we have alignment on that we want to pursue a controller-pased, sync approach for webhook triggers instead.
But as long as we are not there yet (and it's still unclear how to prioritize that work) it's important to me that we have the ability to quickly and confidently react to incidents like the one we had the other week.

Reply:
I think having a default, which can be overriden for specific clone URLs makes perfect sense. It exists largely for the case where you get hit hard, system pages you and you want to adjust the config only to mitigate the issue. But adjusting for all is outside of the scope of that incident and would require deeper investigation.